File: lib/src/capabilities.dart
import 'adapters.dart';

/// Check `capabilities` / `accepted_delta_algos` follow attached adapters (D13).
///
/// `Client.check` defaults to `full_package` only. Extra bits are sent when a
/// live `ArchiveUnpacker` / writable `FileStore` / `Patcher` is attached
/// (typically via `Updater`). Never advertise `binary_delta` without a Patcher.
const capabilityFullPackage = 'full_package';
const capabilityPatchPackage = 'patch_package';
const capabilityFileList = 'file_list';
const capabilityBinaryDelta = 'binary_delta';

const deltaAlgoHdiffpatch = 'hdiffpatch';
const deltaAlgoBsdiff = 'bsdiff';
const deltaAlgoXdelta3 = 'xdelta3';

const telemetryDownloading = 'downloading';
const telemetryApplying = 'applying';
const telemetryInstalled = 'installed';
const telemetryFailed = 'failed';
const telemetryRolledBack = 'rolled_back';

/// D13: never advertise a capability the live adapters cannot perform.
List<String> deriveCapabilities({
  ArchiveUnpacker? unpacker,
  FileStore? fileStore,
  Patcher? patcher,
}) {
  final caps = <String>[capabilityFullPackage];
  if (unpacker != null) {
    caps.add(capabilityPatchPackage);
  }
  if (fileStore != null && fileStore.canWriteFiles) {
    caps.add(capabilityFileList);
  }
  if (deriveDeltaAlgos(patcher).isNotEmpty) {
    caps.add(capabilityBinaryDelta);
  }
  return caps;
}

List<String> deriveDeltaAlgos(Patcher? patcher) {
  if (patcher == null) {
    return const [];
  }
  return [
    for (final algo in patcher.supportedAlgos)
      if (algo.trim().isNotEmpty) algo.trim(),
  ];
}

File: src/capabilities.ts
import type { ArchiveUnpacker, FileStore, Patcher } from "./adapters.js";
import {
  CAPABILITY_BINARY_DELTA,
  CAPABILITY_FILE_LIST,
  CAPABILITY_FULL_PACKAGE,
  CAPABILITY_PATCH_PACKAGE,
} from "./types.js";

export interface CapabilityAdapters {
  unpacker?: ArchiveUnpacker | null;
  fileStore?: FileStore | null;
  patcher?: Patcher | null;
}

/**
 * D13: never advertise a capability the live adapters cannot perform.
 * Default check (Client without extras) stays `full_package` only.
 */
export function deriveCapabilities(adapters: CapabilityAdapters): {
  capabilities: string[];
  accepted_delta_algos?: string[];
} {
  const capabilities = [CAPABILITY_FULL_PACKAGE];
  if (adapters.unpacker) {
    capabilities.push(CAPABILITY_PATCH_PACKAGE);
  }
  if (adapters.fileStore?.canWriteFiles()) {
    capabilities.push(CAPABILITY_FILE_LIST);
  }
  const algos = (adapters.patcher?.supportedAlgos() ?? []).filter((a) => a.length > 0);
  if (algos.length > 0) {
    capabilities.push(CAPABILITY_BINARY_DELTA);
    return { capabilities, accepted_delta_algos: algos };
  }
  return { capabilities };
}
